File: data/code_pairs/fortran/2444.f90
module linked_node_mod
  implicit none
  private
  public :: linked_node

  type, abstract :: linked_node
    integer :: id = 0
  end type linked_node

end module linked_node_mod

module bidir_node_mod
  use linked_node_mod, only: linked_node
  implicit none
  private
  public :: bidir_node

  type, extends(linked_node) :: bidir_node
    class(bidir_node), pointer :: prev => null()
  contains
    procedure :: has_prev
    procedure :: get_prev
    procedure :: set_prev
    procedure :: unset_prev
  end type bidir_node

contains

  elemental logical function has_prev(this)
    class(bidir_node), intent(in) :: this
    has_prev = associated(this%prev)
  end function has_prev

  function get_prev(this) result(res)
    class(bidir_node), intent(in) :: this
    class(bidir_node), pointer :: res
    if (this%has_prev()) then
      res => this%prev
    else
      nullify(res)
    end if
  end function get_prev

  subroutine set_prev(this, new_prev, deallocate_old)
    class(bidir_node), intent(inout) :: this
    class(bidir_node), pointer, intent(in) :: new_prev
    logical, optional, intent(in) :: deallocate_old
    call this%unset_prev(deallocate_old)
    this%prev => new_prev
  end subroutine set_prev

  subroutine unset_prev(this, deallocate_old)
    class(bidir_node), intent(inout) :: this
    logical, optional, intent(in) :: deallocate_old
    if (.not. this%has_prev()) return
    if (present(deallocate_old)) then
      if (deallocate_old) then
        if (associated(this%prev)) then
          call this%prev%unset_prev(.true.)
          deallocate(this%prev)
        end if
        return
      end if
    end if
    nullify(this%prev)
  end subroutine unset_prev

end module bidir_node_mod

File: data/code_pairs/cpp/2444.cpp
#include <iostream>
#include <cassert>
#include <memory>

class LinkedNode {
public:
    int id = 0; // Adding ID for testing purpose as in Fortran code
    virtual ~LinkedNode() = default;
};

class BidirNode : public LinkedNode {
private:
    std::shared_ptr<BidirNode> prev;

public:
    BidirNode() : prev(nullptr) {}

    bool has_prev() const {
        return prev != nullptr;
    }

    std::shared_ptr<BidirNode> get_prev() const {
        return prev; // Direct return works due to smart pointers
    }

    void set_prev(const std::shared_ptr<BidirNode>& new_prev, bool deallocate_old = false) {
        if(deallocate_old && prev) {
            prev.reset(); // Implicit deallocation, custom logic might be needed for deep unset
        }
        prev = new_prev;
    }

    void unset_prev(bool deallocate_old = false) {
        if (deallocate_old && prev) {
            prev.reset(); // Implicit deallocation, custom logic might be needed for deep unset
        } else {
            prev.reset();
        }
    }
};

void test_bidir_node() {
    auto node1 = std::make_shared<BidirNode>();
    auto node2 = std::make_shared<BidirNode>();

    node1->id = 1;
    node2->id = 2;

    std::cout << "Node 1 ID: " << node1->id << std::endl;
    std::cout << "Node 2 ID: " << node2->id << std::endl;

    node1->set_prev(node2);
    std::cout << "Does Node 1 have a previous node? " << (node1->has_prev() ? "true" : "false") << std::endl;
    std::cout << "Node 1's previous node ID: " << node1->get_prev()->id << std::endl;

    node1->unset_prev();
    std::cout << "Does Node 1 have a previous node after unset? " << (node1->has_prev() ? "true" : "false") << std::endl;
}

int main() {
    test_bidir_node();
    return 0;
}